components/core/src/clp/ffi/SchemaTree.hpp

```diff
@@ -0,0 +1,147 @@
+#ifndef CLP_FFI_SCHEMATREE_HPP
+#define CLP_FFI_SCHEMATREE_HPP
+
+#include <cstddef>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include "../ErrorCode.hpp"
+#include "../TraceableException.hpp"
+#include "SchemaTreeNode.hpp"
+
+namespace clp::ffi {
+/**
+ * This class implements a schema tree, providing all necessary methods for upper-layer APIs through
+ * FFI. Nodes are stored using a flattened vector.
+ */
+class SchemaTree {
+public:
+    class OperationFailed : public TraceableException {
+    public:
+        OperationFailed(
+                ErrorCode error_code,
+                char const* const filename,
+                int line_number,
+                std::string message
+        )
+                : TraceableException{error_code, filename, line_number},
+                  m_message{std::move(message)} {}
+
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return m_message.c_str();
+        }
+
+    private:
+        std::string m_message;
+    };
+
+    /**
+     * When locating a tree node, might not always have the tree node id. Instead, we usually use
+     * the parent id, the key name, and the node type to locate a unique tree node. This class wraps
+     * the location information as a non-integer identifier to locate a unique node in the tree.
+     */
+    class TreeNodeLocator {
```

Comment:
This comment could probably be rephrased to be more clear. Maybe something like

"When constructing the schema tree we uniquely identify the location of a node being appended to the try by the unique triple of parent id, key name, and node type. This class
stores that triple, and can act as a unique identifier for a node in the tree."

Reply:
1. How about "appended" -> "inserted"? "Appended" is more implementation-specific, and the doc string doesn't necessarily expose this detail.
2. Shall we add a sentence to explain why the triple is unique? Essentially, it's because key name + node type should not have any ambiguity for a parent node

Reply:
Those both sound good to me.

Reply:
pushed